brave/src/main/java/brave/sampler/RateLimitingSampler.java

```diff
@@ -0,0 +1,103 @@
+package brave.sampler;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Inspired by on https://github.com/aws/aws-xray-sdk-java/blob/2.0.1/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/strategy/sampling/reservoir/Reservoir.java
+ *
+ * Uses only Java 6 APIs to be compatible with Brave Core requirements. Accounts for nanoTime()
+ * overflow, as unlikely as that might be.
+ */
+public class RateLimitingSampler extends Sampler {
+  public static Sampler create(int tracesPerSecond) {
+    if (tracesPerSecond < 0) throw new IllegalArgumentException("tracesPerSecond < 0");
+    if (tracesPerSecond == 0) return Sampler.NEVER_SAMPLE;
+    return new RateLimitingSampler(tracesPerSecond);
+  }
+
+  static final long NANOS_PER_SECOND = TimeUnit.SECONDS.toNanos(1);
+  static final int NANOS_PER_DECISECOND = (int) (NANOS_PER_SECOND / 10);
+
+  final MaxFunction maxFunction;
+  final AtomicInteger usage = new AtomicInteger(0);
+  final AtomicLong nextReset;
+
+  RateLimitingSampler(int tracesPerSecond) {
```

Comment:
@adriancole personally I may want this variable to be reset-able, which mean that we could dynamically change it from outside. We may just need an `apply` function which take new tracesPerSecond and replace maxFunction with new one, HDYT?

Reply:
as mentioned, we'll do this externally